src/sync/arc.rs

```diff
@@ -5,36 +5,72 @@ use std::{mem, ops};
 
 /// Mock implementation of `std::sync::Arc`.
 #[derive(Debug)]
-pub struct Arc<T> {
-    inner: std::sync::Arc<Inner<T>>,
-}
-
-#[derive(Debug)]
-#[repr(C)]
-struct Inner<T> {
-    // This must be the first field to make into_raw / from_raw work
-    value: T,
-
-    obj: rt::Arc,
+pub struct Arc<T: ?Sized> {
+    obj: std::sync::Arc<rt::Arc>,
+    inner: std::sync::Arc<T>,
 }
 
 impl<T> Arc<T> {
     /// Constructs a new `Arc<T>`.
     #[track_caller]
     pub fn new(value: T) -> Arc<T> {
-        let inner = std::sync::Arc::new(Inner {
-            value,
-            obj: rt::Arc::new(location!()),
-        });
+        let std = std::sync::Arc::new(value);
 
-        Arc { inner }
+        Arc::from_std(std)
     }
 
     /// Constructs a new `Pin<Arc<T>>`.
     pub fn pin(data: T) -> Pin<Arc<T>> {
         unsafe { Pin::new_unchecked(Arc::new(data)) }
     }
 
+    /// Returns the inner value, if the `Arc` has exactly one strong reference.
+    pub fn try_unwrap(_this: Arc<T>) -> Result<T, Arc<T>> {
+        unimplemented!();
+    }
+}
+
+impl<T: ?Sized> Arc<T> {
+    /// Converts `std::sync::Arc` to `loom::sync::Arc`
+    ///
+    /// This is needed to create `Arc<T>` where `T: !Sized`
+    ///
+    /// ## Panics
+    ///
+    /// If the provided `Arc` has copies (i.e., if it is not unique).
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// use loom::sync::Arc;
+    ///
+    /// # loom::model::model(|| {
+    /// // std's arc can be automatically coerced
+    /// let std = std::sync::Arc::new([1, 2, 3]);
+    /// let loom: Arc<[u8]> = Arc::from_std(std);
+    ///
+    /// let std = std::sync::Arc::new([1, 2, 3]);
+    /// let loom: Arc<dyn Sync + Send> = Arc::from_std(std);
+    /// # });
```

Comment:
a very minor docs suggestion, take it or leave it: it could also be nice to have a `compile_fail` example showing how `loom::sync::Arc::new` _cannot_ be used with `!Sized` values directly, so that users can see why this function is necessary? not a blocker.